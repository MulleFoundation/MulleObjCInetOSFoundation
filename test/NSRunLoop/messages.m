//
//  main.m
//  archiver-test
//
//  Created by Nat! on 19.04.16.
//  Copyright Â© 2016 Mulle kybernetiK. All rights reserved.
//
#ifdef __MULLE_OBJC__
# import <MulleObjCOSFoundation/MulleObjCOSFoundation.h>
#else
# import <Foundation/Foundation.h>
#endif


//#import "MulleStandaloneObjCFoundation.h"
@interface Foo : NSObject
@end


@implementation Foo

- (void) a:(NSNumber *) nr
{
   printf( "%s %d\n", __PRETTY_FUNCTION__, [nr intValue]);
}


- (void) b:(NSNumber *) nr
{
   printf( "%s %d\n", __PRETTY_FUNCTION__, [nr intValue]);
}


- (void) c:(NSNumber *) nr
{
   printf( "%s %d\n", __PRETTY_FUNCTION__, [nr intValue]);
}


@end



int   main( int argc, const char * argv[])
{
   Foo               *foo;
   id                argument;
   int               i;
   NSDate            *date;
   NSRunLoop         *runLoop;
   NSTimeInterval    interval;
   NSTimeInterval    now;

#ifdef __MULLE_OBJC__
   if( mulle_objc_global_check_universe( __MULLE_OBJC_UNIVERSENAME__) != mulle_objc_universe_is_ok)
      return( 1);
#endif
   foo      = [[Foo new] autorelease];
   argument = [NSNumber numberWithInt:1848];
   runLoop  = [NSRunLoop currentRunLoop];

   [runLoop performSelector:@selector( a:)
                     target:foo
                   argument:argument
                      order:1000
                      modes:@[ @"bar"]];

   [runLoop performSelector:@selector( a:)
                     target:foo
                   argument:argument
                      order:100000
                      modes:@[ @"foo"]];

   [runLoop performSelector:@selector( b:)
                     target:foo
                   argument:argument
                      order:10000
                      modes:@[ @"foo"]];

   [runLoop performSelector:@selector( c:)
                     target:foo
                   argument:argument
                      order:1000
                      modes:@[ @"foo"]];

   // test lots of messages, just to make sure realloc works
   for( i = 0; i < 32; i++)
   {
      [runLoop performSelector:@selector( c:)
                        target:foo
                      argument:[NSNumber numberWithInt:i]
                         order:i
                         modes:@[ @"foo"]];
   }

   //
   // these will eventually produce duplicate orders
   // the earlier messages should be called first if the order is same
   // which is for those on the range 16-48
   //
   for( i = 32; i < 64; i++)
   {
      [runLoop performSelector:@selector( c:)
                        target:foo
                      argument:[NSNumber numberWithInt:i]
                         order:64+16-i
                         modes:@[ @"foo"]];
   }

   [runLoop cancelPerformSelector:@selector( b:)
                           target:foo
                         argument:argument];

   [runLoop _sendMessagesOfRunLoopMode:[runLoop mulleRunLoopModeForMode:@"foo"]];

   return( 0);
}
