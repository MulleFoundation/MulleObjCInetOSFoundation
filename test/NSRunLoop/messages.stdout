-[Foo c:] 0
-[Foo c:] 1
-[Foo c:] 2
-[Foo c:] 3
-[Foo c:] 4
-[Foo c:] 5
-[Foo c:] 6
-[Foo c:] 7
-[Foo c:] 8
-[Foo c:] 9
-[Foo c:] 10
-[Foo c:] 11
-[Foo c:] 12
-[Foo c:] 13
-[Foo c:] 14
-[Foo c:] 15
-[Foo c:] 16
-[Foo c:] 17
-[Foo c:] 63
-[Foo c:] 18
-[Foo c:] 62
-[Foo c:] 19
-[Foo c:] 61
-[Foo c:] 20
-[Foo c:] 60
-[Foo c:] 21
-[Foo c:] 59
-[Foo c:] 22
-[Foo c:] 58
-[Foo c:] 23
-[Foo c:] 57
-[Foo c:] 24
-[Foo c:] 56
-[Foo c:] 25
-[Foo c:] 55
-[Foo c:] 26
-[Foo c:] 54
-[Foo c:] 27
-[Foo c:] 53
-[Foo c:] 28
-[Foo c:] 52
-[Foo c:] 29
-[Foo c:] 51
-[Foo c:] 30
-[Foo c:] 50
-[Foo c:] 31
-[Foo c:] 49
-[Foo c:] 48
-[Foo c:] 47
-[Foo c:] 46
-[Foo c:] 45
-[Foo c:] 44
-[Foo c:] 43
-[Foo c:] 42
-[Foo c:] 41
-[Foo c:] 40
-[Foo c:] 39
-[Foo c:] 38
-[Foo c:] 37
-[Foo c:] 36
-[Foo c:] 35
-[Foo c:] 34
-[Foo c:] 33
-[Foo c:] 32
-[Foo c:] 1848
-[Foo a:] 1848
